zcash_primitives/src/extensions/transparent.rs

```diff
@@ -1,3 +1,163 @@
-//! Zcash transparent extensions.
+//! Core traits and structs for Transparent Zcash Extensions.
 
-pub mod demo;
+use crate::transaction::components::{Amount, OutPoint, TzeOut};
+use std::fmt;
+
+pub trait FromPayload: Sized {
+    type Error;
+
+    /// Parses an extension type from a mode and payload.
+    fn from_payload(mode: usize, payload: &[u8]) -> Result<Self, Self::Error>;
+}
+
+pub trait ToPayload {
+    /// Returns a serialized payload and its corresponding mode.
+    fn to_payload(&self) -> (usize, Vec<u8>);
+}
+
+/// A condition that can be used to encumber transparent funds.
+#[derive(Clone, Debug)]
+pub struct Precondition {
+    pub extension_id: usize,
+    pub mode: usize,
+    pub payload: Vec<u8>,
+}
+
+impl Precondition {
+    pub fn from<P: ToPayload>(extension_id: usize, value: &P) -> Precondition {
+        let (mode, payload) = value.to_payload();
+        Precondition {
+            extension_id,
+            mode,
+            payload,
+        }
+    }
+
+    pub fn try_to<P: FromPayload>(&self) -> Result<P, P::Error> {
+        P::from_payload(self.mode, &self.payload)
+    }
+}
```

Comment:
Where is this used? If it is necessary, this could probably be implemented using the standard traits:
```suggestion
}

use std::convert::TryFrom;

impl<'a, P: FromPayload> TryFrom<&'a Precondition> for P {
    type Error = P::Error;

    fn try_from(p: &'a Precondition) -> Result<Self, Self::Error> {
        P::from_payload(p.mode, &p.payload)
    }
}
```

Reply:
This seems to fall afoul of the error described in https://stackoverflow.com/a/32431031/390636 - I've spent a bit trying to figure it out, but to no avail, suggestions?

Reply:
For reference, 
```
error[E0119]: conflicting implementations of trait std::convert::TryFrom<&extensions::transparent::Precondition>:
  --> zcash_primitives/src/extensions/transparent.rs:38:1
   |
38 | impl<'a, P: FromPayload> TryFrom<&'a Precondition> for P {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T, U> std::convert::TryFrom<U> for T
             where U: std::convert::Into<T>;

error[E0210]: type parameter `P` must be used as the type parameter for some local type (e.g., `MyStruct<P>`)
  --> zcash_primitives/src/extensions/transparent.rs:38:1
   |
38 | impl<'a, P: FromPayload> TryFrom<&'a Precondition> for P {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type parameter `P` must be used as the type parameter for some local type
   |
   = note: only traits defined in the current crate can be implemented for a type parameter

error: aborting due to 2 previous errors
```